LLama/LLamaTemplate.cs

```diff
@@ -0,0 +1,301 @@
+    private bool _addAssistant;
+
+    /// <summary>
+    /// Temporary array of messages in the format llama.cpp needs, used when applying the template
+    /// </summary>
+    private LLamaChatMessage[] _nativeChatMessages = new LLamaChatMessage[4];
+
+    /// <summary>
+    /// Indicates how many bytes are in <see cref="_result"/> array
+    /// </summary>
+    private int _resultLength;
+
+    /// <summary>
+    /// Result bytes of last call to <see cref="Apply"/>
+    /// </summary>
+    private byte[] _result = Array.Empty<byte>();
+
+    /// <summary>
+    /// Indicates if this template has been modified and needs regenerating
+    /// </summary>
+    private bool _dirty = true;
+    #endregion
+
+    #region properties
+    /// <summary>
+    /// Number of messages added to this template
+    /// </summary>
+    public int Count { get; private set; }
+
+    /// <summary>
+    /// Get the message at the given index
+    /// </summary>
+    /// <param name="index"></param>
+    /// <returns></returns>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown if index is less than zero or greater than or equal to <see cref="Count"/></exception>
+    public (string role, string content) this[int index]
+    {
+        get
+        {
+            if (index < 0)
+                throw new ArgumentOutOfRangeException(nameof(index), "Index must be >= 0");
+            if (index >= Count)
+                throw new ArgumentOutOfRangeException(nameof(index), "Index must be < Count");
+
+            return (_messages[index].Role, _messages[index].Content);
+        }
+    }
+
+    /// <summary>
+    /// Whether to end the prompt with the token(s) that indicate the start of an assistant message.
+    /// </summary>
+    public bool AddAssistant
+    {
+        get => _addAssistant;
+        set
+        {
+            if (value != _addAssistant)
+            {
+                _dirty = true;
+                _addAssistant = value;
+            }
+        }
+    }
+    #endregion
+
+    #region construction
+    /// <summary>
+    /// Construct a new template, using the default model template
+    /// </summary>
+    /// <param name="model"></param>
+    public LLamaTemplate(SafeLlamaModelHandle model)
+    {
+        _model = model;
+    }
+
+    /// <summary>
+    /// Construct a new template, using the default model template
+    /// </summary>
+    /// <param name="weights"></param>
+    public LLamaTemplate(LLamaWeights weights)
+        : this(weights.NativeHandle)
+    {
+    }
+
+    /// <summary>
+    /// Construct a new template, using a custom template.
+    /// </summary>
+    /// <remarks>Only support a pre-defined list of templates. See more: https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template</remarks>
+    /// <param name="customTemplate"></param>
+    public LLamaTemplate(string customTemplate)
+    {
+        _customTemplate = Encoding.UTF8.GetBytes(customTemplate + "\0");
```

Comment:
In this case the encoding should be whatever llama.cpp expects, since these bytes will be passed directly to the the `llama_chat_apply_template ` function. I think that's UTF8?

Reply:
I haven't looked into the template of llama.cpp yet. I'm just not sure if other encodings are required when using other languages (for example the default encoding for Chinese on Windows is `GB2312`). Maybe we could add a property `Encoding` with default value `Encoding.UTF8` to `LLamaTemplate` and use it here and in [this place](https://github.com/SciSharp/LLamaSharp/pull/715/files#diff-c399c5472f555b08aeef846334552d34feb2361ea8e417afe86976721536d118R306)? We could hide it from users now, setting it as `private readonly`. Thus it will be easy for us to add support for multi-encoding in the future, if that's required.

Reply:
I've added a `private static readonly Encoding` field.